client/widgets/CArtifactHolder.cpp

```diff
@@ -945,3 +943,102 @@ bool CArtifactsOfHero::SCommonPart::Artpos::valid()
 	assert(AOH && art);
 	return art == AOH->curHero->getArt(slotID);
 }
+
+CArtPlace::CArtPlace(Point position, const CArtifactInstance * Art) : ourArt(Art)
+{
+	pos += position;
+	pos.w = pos.h = 44;
+}
+
+void CArtPlace::clickLeft(tribool down, bool previousState)
+{
+	LRClickableAreaWTextComp::clickLeft(down, previousState);
+}
+
+void CArtPlace::clickRight(tribool down, bool previousState)
+{
+	LRClickableAreaWTextComp::clickRight(down, previousState);
+}
+
+CCommanderArtPlace::CCommanderArtPlace(Point position, const CGHeroInstance * commanderOwner, ArtifactPosition artSlot, const CArtifactInstance * Art) : CArtPlace(position, Art), commanderOwner(commanderOwner), commanderSlotID(artSlot.num)
+{
+	createImage();
+	setArtifact(Art);
+}
+
+void CCommanderArtPlace::clickLeft(tribool down, bool previousState)
+{
+	if (down && ourArt && text.size())
+		LOCPLINT->showYesNoDialog("Do you want to give this artifact back to hero?", [this] { returnArtToHeroCallback(); }, [] {});
+		//CArtPlace::clickLeft(down, previousState);	
+}
+
+void CCommanderArtPlace::clickRight(tribool down, bool previousState)
+{
+	if (down && ourArt && text.size())
+		CArtPlace::clickRight(down, previousState);
+}
+
+void CCommanderArtPlace::createImage()
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL;
+
+	int graphic = 0;
+	if (ourArt)
+		graphic = ourArt->artType->iconIndex;
+
+	image = new CAnimImage("artifact", graphic);
+	if (!ourArt)
+		image->disable();
+}
+
+void CCommanderArtPlace::returnArtToHeroCallback()
+{
+	ArtifactPosition artifactPos = commanderSlotID;;
+	ArtifactPosition freeSlot = ourArt->firstBackpackSlot(commanderOwner);
+
+	ArtifactLocation src(commanderOwner->commander.get(), artifactPos);
+	ArtifactLocation dst(commanderOwner, freeSlot);
+
+	if (ourArt->canBePutAt(dst, true))
+	{
+		LOCPLINT->cb->swapArtifacts(src, dst);
+		setArtifact(nullptr);
+		parent->redraw();
```

Comment:
This line is probably a code smell, but I did not find better solution with current design.